### PR TITLE
2.5.x - 	Replaced deprecated Cache access with CacheApi

### PR DESCRIPTION
**Pull request by pr0master**
Replaced deprecated Cache access with CacheApi which supports dependency injection

### CODE REVIEW

File: src/main/java/org/pac4j/play/store/PlaySessionStore.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package org.pac4j.play.store;
+
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.play.PlayWebContext;
+
+/**
+ * Created by ie01034 on 17-6-2016.
```
Comment (leleuj) [inline src/main/java/org/pac4j/play/store/PlaySessionStore.java]: Please type in a real Javadoc, `@author`, `@version`...

**Comment by leleuj** [comment]
It's an excellent work. Thanks for your contribution.

This looks good to me. Can you update the demos: https://github.com/pac4j/play-pac4j-scala-demo and https://github.com/pac4j/play-pac4j-java-demo as well?

This PR is out-of-sync, certainly because of https://github.com/pac4j/play-pac4j/pull/118, you need to merge with master.

I'm hesitating about the new version 2.5.0-SNAPSHOT, this could fit in 2.4.0...

File: src/main/java/org/pac4j/play/store/PlaySessionStore.java (hunk 1 above)
Comment (pr0master) [inline src/main/java/org/pac4j/play/store/PlaySessionStore.java]: I'll do that of course.

**Comment by wwbakker** [comment]
Hi leleuj,

I'm a colleague of pr0master. The reason why we propose to change the major version is because it contains breaking changes for the projects that implement it, as described in the migration guide in the readme here:

https://github.com/pr0master/play-pac4j/tree/2.5.x#240-play-25---250-play-25

On the other hand, the 2.5.0 is still in SNAPSHOT, so I suppose it could still be appropriate to keep it 2.4.0-SNAPSHOT...

File: src/main/scala/org/pac4j/play/filters/SecurityFilter.scala [hunk 2]
```diff
@@ -33,63 +35,74 @@ import scala.compat.java8.FutureConverters._
   * `_anonymous_` will disable authentication and authorization for urls matching the regex.
   * `_authenticated_` will require authentication, but will set clients and authorizers both to `null`.
   *
-  * Rules are applied top to bottom. The first matching rule will define which clients and authorizers
+  * Rules are traversed top to bottom. The first matching rule will define which clients and authorizers
```
Comment (leleuj) [inline src/main/scala/org/pac4j/play/filters/SecurityFilter.scala]: so "traversed and applied"

**Comment by leleuj** [comment]
I just did a new pass on this pull request: see my few comments, no real issue. This looks good.

I think you're right about the versioning, I'll merge that after the 2.4 release in a new 2.5 version.

Can you prepare the PRs for: https://github.com/pac4j/play-pac4j-scala-demo and https://github.com/pac4j/play-pac4j-java-demo to match these changes?

Thanks

**Comment by wwbakker** [comment]
Last friday, pr0master and I were busy implementing the changes into the demo projects, and it was going well for the most part. The problem we are facing however is that SecurityModule.scala / java contains the following piece of example code:

` // CAS
    val casClient = new CasClient("https://casserverpac4j.herokuapp.com/login")
    casClient.setLogoutHandler(**new PlayCacheLogoutHandler()**)
    casClient.setCasProtocol(CasProtocol.CAS20)
`

The logout handler is instantiated inside the module directly, however it now requires a CacheApi instance which is only available through dependency injection (and therefore, only available after the code in the SecurityModule is run). 

Fixing this would also require refactoring the CAS client in pac4j core. Furthermore it would not be possible to add it to the Config object from within the SecurityModule. This is a hard thing to fix for us, since we don't have any experience with CAS.

We can just remove CAS support in the demo for now and fix the rest of the demo for the rest of the clients if you want? I think in the end it is inevitable that (something like) this pull request is done, because getting the cache through the singleton is deprecated already in 2.5, so it will be removed eventually. We already had issues with it right now when upgrading from play 2.4 to 2.5 (hence the reason for this pull request).

Note: Pr0master is unavailable from Monday through Wednesday next week, so the earliest we can work on this again is the end of the week.

**Comment by wwbakker** [comment]
For reference, play uses Google's Guice for dependency injection: (https://github.com/google/guice/wiki/GettingStarted)

The root of the problem is that we need the CacheApi, and we can only get it after the module is already loaded.

We'll look into it a bit further.

**Comment by leleuj** [comment]
OK. Keep me posted.

File: README.md [hunk 3]
```diff
@@ -37,7 +38,7 @@ Just follow these easy steps to secure your Play 2 web application:
 
 You need to add a dependency on:
 
-- the `play-pac4j` library (<em>groupId</em>: **org.pac4j**, *version*: **2.4.0**)
+- the `play-pac4j` library (<em>groupId</em>: **org.pac4j**, *version*: **2.4.0-SNAPSHOT**)
```
Comment (leleuj) [inline README.md]: You certainly mean `2.5.0-SNAPSHOT`.

File: README.md [hunk 4]
```diff
@@ -477,7 +478,7 @@ If you have any question, please use the following mailing lists:
 
 ## Development
 
-The version 2.4.1-SNAPSHOT is under development.
+The version 2.4.0-SNAPSHOT is under development.
```
Comment (leleuj) [inline README.md]: You certainly mean `2.5.0-SNAPSHOT`.

**Comment by leleuj** [comment]
I created the 2.4.x branches for the main project and its demos, master is now at version 2.5.0-SNAPSHOT, ready to accept your changes...

How is this issue going on?

**Comment by wwbakker** [comment]
That's excellent. We haven't gotten around to it yet, but it's on our todo list.

**Comment by pr0master** [comment]
I created both for Java (https://github.com/pac4j/play-pac4j-java-demo/pull/37) and Scala(https://github.com/pac4j/play-pac4j-scala-demo/pull/35) demo's a PR. 

As you can see, we used a provider to make DI possible.

File: src/main/java/org/pac4j/play/cas/logout/PlayCacheLogoutHandler.java [hunk 5]
```diff
@@ -21,16 +24,34 @@
...
+    private final Provider<CacheApi> cacheApiProvider;
+
+    @Inject
+    public PlayCacheLogoutHandler(final CacheApi cacheApi) {
+        this.cacheApiProvider = null;
+        this.cache = cacheApi;
+    }
+
+    public PlayCacheLogoutHandler(final Provider<CacheApi> cacheApiProvider) {
+        this.cache = null;
```
Comment (leleuj) [inline src/main/java/org/pac4j/play/cas/logout/PlayCacheLogoutHandler.java]: In fact, I would not create a `cacheApiProvider` property and directly assign the `cache` property from the provider.

File: src/main/java/org/pac4j/play/store/PlaySessionStore.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package org.pac4j.play.store;
+
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.play.PlayWebContext;
+
+/**
+ * To store data in session.
+ * Extending the SessionStore is necessary for dependency injection to work.
+ * @author Ilias Elkhalloufi and Wessel Bakker
+ * @since 1.9.0
```
Comment (leleuj) [inline src/main/java/org/pac4j/play/store/PlaySessionStore.java]: You mean `2.5.0`?

**Comment by leleuj** [comment]
Two minor comments, but everything looks good to me. Anything left on your side?

File: src/main/java/org/pac4j/play/cas/logout/PlayCacheLogoutHandler.java (hunk 5 above)
Comment (wwbakker) [inline src/main/java/org/pac4j/play/cas/logout/PlayCacheLogoutHandler.java]: Sadly this does not work:
The PlayCacheLogoutHandler instantiated manually in SecurityModule.configure(). At that time, there not has been initialisation or dependency injection by guice. 

After SecurityModule.configure() is run, the application is started and the CacheApi becomes available.
Only then Provider<CacheApi> will produce a value.

Comment (leleuj) [inline src/main/java/org/pac4j/play/cas/logout/PlayCacheLogoutHandler.java]: OK

**Comment by pr0master** [comment]
Changed the @since attribute for correction. We have nothing to add. Good to go!

**Comment by leleuj** [comment]
Merging...

**Comment by leleuj** [comment]
@wwbakker @ielkhalloufi I'm generally satisfied by this change, but I'd like to revisit two things:

1) I plan to rename the `PlayCacheStore` as the `PlayCacheSessionStore`. In the future, we will have two concepts in *pac4j*: the `Store` (an abstraction for a cache) and the `SessionStore` (an abstraction for the web session). So I'd like to properly distinguish between these two concepts with appropriate names

2) I'd like to get rid of the `PlaySessionStore` interface: Play already offers the capacity to switch from cache, but *pac4j* will go further by allowing users to switch from web session implementation. For that, we don't need to constraint to a specific `PlaySessionStore`, but only to a generic `SessionStore<PlayWebContext>`. Does it make sense or feel overkill?

**Comment by millross** [comment]
I don’t understand the need for pac4j to switch web session implementations. Surely selection of web session implementation should be a concern for the web framework, not the authentication/authorization framework.

Certainly vertx-web (which underlies the vertx-pac4j implementation) already allows selection of session store implementation (and creation of new ones), and this is presumably true for other web frameworks too. Where’s the benefit to the pac4j user in pac4j also implementing this (which creates extra maintenance overhead on our side in terms of mapping an existing implementation to a new abstraction)? Generally someone using a web framework will want to use the mechanisms associated with that framework for selection of session implementation, as that is what they are likely to be familiar with, so what are the likely scenarios when we would expect the web framework to use one session implementation and pac4j to use another? If we force pac4j users to actually have to configure to use the session store they’ve selected for their web framework, then I think that’s an unnecessary friction, and if we don’t force it, I doubt the option will be of interest to anyone. 

So my question is: who has asked for the ability to use a different session store impl from the one they’ve chosen for their web? In the case of vert.x this will likely lead to a user model completely incompatible with the vert.x one, which means there will be no point in a vert.x impl of this, because noone will want to use it.

**Comment by leleuj** [comment]
OK. I understand you answer to point 2) is no ;-)

*pac4j* needs to interact with the web session to save / restore user profiles. So we really need a `SessionStore` concept which represents these abstractions. And we need go a little further in this concept to be able to handle the CAS logout for example without a `CasLogoutHandler` for each framework.

But you're right, I'm loosing myself here, switching from web session implementations should be done at the web framework level, not at the *pac4j* level. Specific session stores should only delegate to the underlying web framework.

**Comment by millross** [comment]
By the way, I’m completely cool with the idea of having a standard interface for that web session interaction (so that maybe some more logic can be pushed into core). My only issue is the idea of pluggable session store implementations for pac4j whereby we might end up using different session store mechanism between the web framework and pac4j, which I see as the first step on a road to hell.

The same can be applied for shared caches for example (e.g. vert.x already has a shared memory abstraction). I just want to try and ensure that there’s one implementor of each interface per web framework, not one per storage technology (redis, hazelcast, for example). Does that make sense? In other words, a common interface to be accessed by core logic makes sense to me (and would ease framework implementation in my view), it’s just at what level we choose to implement that interface that was my concern :)

**Comment by leleuj** [comment]
You're right: we should not have different store mechanisms between *pac4j* and the web framework and the *pac4j*  `SessionStore` should rely on the web framework mechanism (that's the only solution to have one mechanism at the end).

About the shared caches, I'm not sure to follow you.

**Comment by wwbakker** [comment]
Unfortunately removing PlaySessionStore will not work: The reason we
introduced it is because dependency injection works only with actual
classes/interfaces (e.g. PlaySessionStore), and not classes/interfaces with
a generic type (e.g. SessionStore<T>).

**Comment by leleuj** [comment]
In fact, with Guice, you can use the `TypeLiteral` (https://google.github.io/guice/api-docs/latest/javadoc/index.html?com/google/inject/TypeLiteral.html). But this  `PlaySessionStore` makes sense.

**Comment by wwbakker** [comment]
Interessant, thanks

Op 10 sep. 2016 9:53 AM schreef "LELEU Jérôme" <notifications@github.com>:

**Comment by wwbakker** [comment]
Interesting*. Sorry, autocorrect "fixed" it.

Op 10 sep. 2016 1:32 PM schreef "Wessel W. Bakker" <wwbakker@gmail.com>:

**Pull request merged**